### PR TITLE
Improved ball search

### PR DESCRIPTION
**Pull request by JonasHolz**
## Introduced Changes

Add a new note that outputs the most likely ball_position (suggested_ball_position) from a heat map, if the suggested_ball_position has a high enough validity
Change search behavior 
- After loosing the ball the offensive robots will now walk to the suggested_ball_position
- If there is no suggested_ball_position the robot will go back to the old search behavior
Add a heatmap layer in Twix 

## ToDo / Known Issues

- Testing of robot behavior with real robots in a test game
- Maybe some parameter changes

## Ideas for Next Iterations (Not This PR)

- Improve twix heatmap visual design
- Implement more changes to search behavior
- Lower heatmap validity for checked areas
- Set hypothesis for corner kicks, kick ins, etc.
- Better fake data for behavior-simulator

## How to Test

Run ball search Lua test files to see the implemented behavior
Let the Robot kick the ball across the field than hide it to see if he still walks to the last seen ball position

### CODE REVIEW

**Comment by MaikRe** [comment]
Looks good so far! I would like to see this in a test game before we merge

File: crates/control/src/search_suggestor.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+use serde::{Deserialize, Serialize};
+use types::{
+    ball_position::{BallPosition, HypotheticalBallPosition},
+    field_dimensions::FieldDimensions,
+    parameters::SearchSuggestorParameters,
+};
+
+#[derive(Deserialize, Serialize)]
+pub struct SearchSuggestor {
+    heatmap_dimensions: (usize, usize),
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:14]: Maybe don't save this in the struct but use `self.heatmap.shape()` instead

File: crates/control/src/search_suggestor.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub suggested_search_position: MainOutput<Option<Point2<f32>>>,
+}
+
+impl SearchSuggestor {
+    pub fn new(_context: CreationContext) -> Result<Self> {
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:41]: remove `_` if used

File: crates/control/src/search_suggestor.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        let heatmap_dimensions = (
+            _context.field_dimensions.length.round() as usize
+                * _context.search_suggestor_configuration.cells_per_meter,
+            _context.field_dimensions.width.round() as usize
+                * _context.search_suggestor_configuration.cells_per_meter,
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:46]: First multiply, then round. This makes the result more correct.
If field-length were 2.4 and you had 10 cells per meter, you would calculate around 20 cells whereas (2.4 * 10).round() returns 24 cells which is what was specified in the config

File: crates/control/src/search_suggestor.rs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub suggested_search_position: MainOutput<Option<Point2<f32>>>,
+}
+
+impl SearchSuggestor {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        let heatmap_dimensions = (
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:42]: use destructuring here, for example`let (number_of_cells_length, number_of_cells_width) = ...`

File: crates/control/src/search_suggestor.rs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        self.update_heatmap(
+            context.ball_position,
+            context.invalid_ball_positions,
+            context.robot_to_field.copied(),
+            context.search_suggestor_configuration.cells_per_meter,
+            context.search_suggestor_configuration.heatmap_decay_factor,
+        );
+        let maximum_heat_heatmap_position = self.heatmap.iamax_full();
+        let mut suggested_search_position: Option<Point2<f32>> = None;
+        if self.heatmap.get(maximum_heat_heatmap_position).is_some() {
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:64]: How can the index of the maximum index be out of bounds? In my understanding this cannot happen. If this is correct, use normal array indexing syntax `[...]`

File: crates/control/src/search_suggestor.rs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+                let mut search_suggestion = Point2::new(
+                    maximum_heat_heatmap_position.0 as f32
+                        / context.search_suggestor_configuration.cells_per_meter as f32,
+                    maximum_heat_heatmap_position.1 as f32
+                        / context.search_suggestor_configuration.cells_per_meter as f32,
+                );
+                let length_half = context.field_dimensions.length / 2.0;
+                let width_half = context.field_dimensions.width / 2.0;
+
+                if search_suggestion.x >= length_half {
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:77]: idea: maybe make a newtype for the heatmap that handles the coordinate conversions for you. That way you primary control flow path becomes cleaner.

File: crates/control/src/search_suggestor.rs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                } else {
+                    search_suggestion.x = length_half - search_suggestion.x
+                }
+                if search_suggestion.y >= width_half {
+                    search_suggestion.y -= width_half;
+                } else {
+                    search_suggestion.y = width_half - search_suggestion.y
+                }
+
+                search_suggestion.x +=
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:88]: What is this for?

File: crates/control/src/search_suggestor.rs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        if let Some(ball_position) = ball_position {
+            if let Some(robot_to_field) = robot_to_field {
+                let ball_heatmap_position = self.calculate_heatmap_position(
+                    cells_per_meter,
+                    robot_to_field * ball_position.position,
+                );
+                if self.heatmap.get(ball_heatmap_position).is_some() {
+                    self.heatmap[ball_heatmap_position] = 1.0;
+                } else {
+                    println!("Invalid ball heatmap position");
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:122]: Remove?

File: crates/control/src/search_suggestor.rs [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+            if let Some(robot_to_field) = robot_to_field {
+                let heatmap_position = self.calculate_heatmap_position(
+                    cells_per_meter,
+                    robot_to_field * ball_hypothesis.position,
+                );
+                if self.heatmap.get(heatmap_position).is_some() {
+                    self.heatmap[heatmap_position] =
+                        (self.heatmap[heatmap_position] + ball_hypothesis.validity) / 2.0;
+                } else {
+                    println!("Invalid hypothesis heatmap position");
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:136]: remove?

File: crates/control/src/search_suggestor.rs [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+                }
+            }
+        }
+        self.heatmap = self.heatmap.clone() * (1.0 - heatmap_decay_factor);
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:140]: Why clone here? You could_use `self.heatmap.scale_mut(1.0 - heatmap_decay_factor)`

File: crates/control/src/search_suggestor.rs [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    fn calculate_heatmap_position(
+        &mut self,
+        cells_per_meter: usize,
+        hypothesis_position: Point2<f32>,
+    ) -> (usize, usize) {
+        let mut x_position: usize = 0;
```
Comment (okiwi6) [inline crates/control/src/search_suggestor.rs:148]: maybe you could create an `Isometry2` for coordinate conversion, apply that and then just convert to usize?

File: crates/serialize_hierarchy/src/not_supported.rs [hunk 12]
```diff
@@ -98,6 +98,7 @@ implement_as_not_supported!(u64);
 implement_as_not_supported!(usize);
 // nalgebra
 implement_as_not_supported!(SMatrix<f32, 3, 3>);
+implement_as_not_supported!(DMatrix<f32>);
```
Comment (okiwi6) [inline crates/serialize_hierarchy/src/not_supported.rs:101]: alphabetical sorting s'il vous plaît

**Review by okiwi6 — CHANGES_REQUESTED** [review]
Thanks for this big PR, some style remarks left :100:

File: crates/control/src/search_suggestor.rs (hunk 1 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 2 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 3 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 4 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 5 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: You are right. I changed it to normal array indexing syntax

File: crates/control/src/search_suggestor.rs (hunk 6 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: I changed the calculate_heatmap_position function via a Similarity such that we can skip this whole calculation, which results in a cleaner control flow.

File: crates/control/src/search_suggestor.rs (hunk 7 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: This part was also removed as stated above

File: crates/control/src/search_suggestor.rs (hunk 8 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 9 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 10 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 11 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: After some consideration, I came to the conclusion that putting the hole heatmap into a type was the best way to deal with the coordinate transformation.

File: crates/control/src/search_suggestor.rs [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+pub struct CreationContext {
+    field_dimensions: Parameter<FieldDimensions, "field_dimensions">,
+    search_suggestor_configuration: Parameter<SearchSuggestorParameters, "search_suggestor">,
+}
+
+#[context]
+pub struct CycleContext {
+    search_suggestor_configuration: Parameter<SearchSuggestorParameters, "search_suggestor">,
+    ball_position: Input<Option<BallPosition<Ground>>, "ball_position?">,
+    invalid_ball_positions: Input<Vec<HypotheticalBallPosition<Ground>>, "invalid_ball_positions">,
```
Comment (oleflb) [inline crates/control/src/search_suggestor.rs:31]: what about `removed_all_positions`

File: crates/control/src/search_suggestor.rs [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        self.update_heatmap(
+            context.ball_position,
+            context.invalid_ball_positions,
+            context.ground_to_field.copied(),
+            context.search_suggestor_configuration.heatmap_decay_factor,
+        );
+        let maximum_heat_heatmap_position = self.heatmap.map.iamax_full();
```
Comment (oleflb) [inline crates/control/src/search_suggestor.rs:68]: extract to function in `HeatMap` struct, should also do the conversion and return `Point<Field>`

File: crates/control/src/search_suggestor.rs [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+#[derive(Deserialize, Serialize)]
+struct Heatmap {
+    map: DMatrix<f32>,
+    field_dimensions: FieldDimensions,
+    cells_per_meter: f32,
+}
+
+impl Heatmap {
+    fn field_to_heatmap(&self, index: Point2<Field>) -> (usize, usize) {
```
Comment (oleflb) [inline crates/control/src/search_suggestor.rs:124]: rename `index` to sth. like `field_point`?

File: crates/control/src/search_suggestor.rs [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+        (
+            clamp(heatmap_point.0, 0, self.map.shape().0 - 1),
+            clamp(heatmap_point.1, 0, self.map.shape().1 - 1),
+        )
+    }
+}
+
+impl Index<Point2<Field>> for Heatmap {
+    type Output = f32;
+    fn index(&self, index: Point2<Field>) -> &Self::Output {
```
Comment (oleflb) [inline crates/control/src/search_suggestor.rs:138]: same here

File: crates/control/src/search_suggestor.rs [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+        let heatmap_point = self.field_to_heatmap(index);
+        &self.map[heatmap_point]
+    }
+}
+
+impl IndexMut<Point2<Field>> for Heatmap {
+    fn index_mut(&mut self, index: Point2<Field>) -> &mut Self::Output {
```
Comment (oleflb) [inline crates/control/src/search_suggestor.rs:145]: same here

File: tests/behavior/ball_search.lua [hunk 18]
```diff
@@ -0,0 +1,75 @@
+local inspect = require 'inspect'
+print("Hello world from lua!")
+
+function spawn_robot(number)
+    table.insert(state.robots, create_robot(number))
+end
+
+-- spawn_robot(1)
+-- spawn_robot(2)
```
Comment (oleflb) [inline tests/behavior/ball_search.lua:9]: commented?

File: tools/twix/src/panels/map/mod.rs [hunk 19]
```diff
@@ -78,6 +78,7 @@ pub struct MapPanel {
     feet_detection: EnabledLayer<layers::FeetDetection, Ground>,
     ball_filter: EnabledLayer<layers::BallFilter, Ground>,
     obstacle_filter: EnabledLayer<layers::ObstacleFilter, Ground>,
+    ball_search_heatmap: EnabledLayer<layers::BallSearchHeatmap, Field>,
```
Comment (oleflb) [inline tools/twix/src/panels/map/mod.rs:81]: is it wise to draw such a large feature last?

**Review by oleflb — CHANGES_REQUESTED** [review]
Looks much cleaner :100:

File: crates/control/src/search_suggestor.rs (hunk 13 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs:31]: These will be relevant for decreasing the validity for checked areas, which is a thing for the next iteration of ball search.

File: crates/control/src/search_suggestor.rs (hunk 14 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 15 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 16 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: crates/control/src/search_suggestor.rs (hunk 17 above)
Comment (JonasHolz) [reply crates/control/src/search_suggestor.rs]: Done

File: tests/behavior/ball_search.lua (hunk 18 above)
Comment (JonasHolz) [reply tests/behavior/ball_search.lua]: I removed the commented lines.

File: tools/twix/src/panels/map/mod.rs (hunk 19 above)
Comment (JonasHolz) [reply tools/twix/src/panels/map/mod.rs]: You are right. I shifted it to be painted after lines.

**Review by oleflb — APPROVED** [review]
:steam_locomotive:

**Pull request merged**